src/session.rs

```diff
@@ -0,0 +1,140 @@
+        } else {
+            let mut path = dirs::config_dir()
+                .context("couldn't get config directory")?
+                .join::<PathBuf>(["xh", "sessions", host].iter().collect());
+            path.push(format!("{}.json", name_or_path));
+            path
+        };
+
+        let content = match fs::read_to_string(&path) {
+            Ok(content) => serde_json::from_str::<Content>(&content)?,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => Content::default(),
+            Err(err) => return Err(err.into()),
+        };
+
+        // TODO: remove expired cookies
+        Ok(Session {
+            path,
+            read_only,
+            content,
+        })
+    }
+
+    pub fn headers(&self) -> Result<HeaderMap> {
+        let mut headers: HeaderMap = HeaderMap::try_from(&self.content.headers)?;
+        let cookies = self
+            .content
+            .cookies
+            .values()
+            .map(|value| Cookie::parse(value).unwrap())
+            .map(|c| format!("{}={}", c.name(), c.value()))
+            .collect::<Vec<_>>()
+            .join("; ");
+        if !cookies.is_empty() {
+            headers.insert(COOKIE, HeaderValue::from_str(&cookies)?);
+        }
+        Ok(headers)
+    }
+
+    pub fn save_headers(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        for (key, value) in request_headers.iter() {
+            let key = key.as_str();
+            if key != "cookie" && !key.starts_with("content-") && !key.starts_with("if-") {
+                self.content
+                    .headers
+                    .insert(key.into(), value.to_str()?.into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn save_auth(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        if let Some(value) = request_headers.get(AUTHORIZATION) {
+            self.content
+                .headers
+                .insert("authorization".into(), value.to_str()?.into());
+        }
+        Ok(())
+    }
+
+    pub fn save_cookies(&mut self, response_headers: &HeaderMap) -> Result<()> {
+        for cookie in response_headers.get_all(SET_COOKIE) {
+            let raw_cookie = cookie.to_str()?;
+            let parsed_cookie = Cookie::parse(raw_cookie)?;
+            self.content
+                .cookies
+                .insert(parsed_cookie.name().into(), raw_cookie.into());
+        }
+        Ok(())
+    }
+
+    pub fn persist(&self) -> Result<()> {
+        if let Some(parent_path) = self.path.parent() {
+            fs::create_dir_all(parent_path)?;
+        }
+        let mut session_file = fs::File::create(&self.path)?;
+        if !self.read_only {
+            let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
+            let mut ser = serde_json::Serializer::with_formatter(&mut session_file, formatter);
+            self.content.serialize(&mut ser)?;
+            session_file.write_all(b"\n")?;
+        }
```

Comment:
I don't think this is right. It can leave blank files.

HTTPie creates a file if `read_only` is false *or* if the file doesn't exist yet. Otherwise it does nothing.

Reply:
oh, I must have misread your earlier comment. So I am assuming we would want something like this:
```rust
if !args.is_session_read_only {
    if let Some(ref mut s) = session {
        s.persist()
            .context(format!("couldn't persist session {:?}", s.path))?;
    }
}
```

Reply:
HTTPie also persists if the session is new, i.e. if the file doesn't exist yet. See [`client.py`](https://github.com/httpie/httpie/blob/fc45bf0fe32711cbbdda2eb5f71ef7cd25b08312/httpie/client.py#L125), [`config.py`](https://github.com/httpie/httpie/blob/fc45bf0fe32711cbbdda2eb5f71ef7cd25b08312/httpie/config.py#L81).